src/com/xilinx/rapidwright/design/tools/InlineFlopTools.java

```diff
@@ -0,0 +1,265 @@
+ * All rights reserved.
+ *
+ * Author: Chris Lavin, AMD Advanced Research and Development.
+ *
+ * This file is part of RapidWright.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.xilinx.rapidwright.design.tools;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.xilinx.rapidwright.design.Cell;
+import com.xilinx.rapidwright.design.Design;
+import com.xilinx.rapidwright.design.DesignTools;
+import com.xilinx.rapidwright.design.Net;
+import com.xilinx.rapidwright.design.SiteInst;
+import com.xilinx.rapidwright.design.SitePinInst;
+import com.xilinx.rapidwright.design.Unisim;
+import com.xilinx.rapidwright.design.blocks.PBlock;
+import com.xilinx.rapidwright.device.BEL;
+import com.xilinx.rapidwright.device.Series;
+import com.xilinx.rapidwright.device.Site;
+import com.xilinx.rapidwright.eco.ECOPlacementHelper;
+import com.xilinx.rapidwright.edif.EDIFCell;
+import com.xilinx.rapidwright.edif.EDIFCellInst;
+import com.xilinx.rapidwright.edif.EDIFHierNet;
+import com.xilinx.rapidwright.edif.EDIFNet;
+import com.xilinx.rapidwright.edif.EDIFPort;
+import com.xilinx.rapidwright.edif.EDIFPortInst;
+import com.xilinx.rapidwright.util.Pair;
+import com.xilinx.rapidwright.util.StringTools;
+
+/**
+ * Set of tools to add/remove flip flops inline to top level port connections.
+ * This is targeted at kernel replication preparation prior to routing to ensure
+ * that connections are routed outside of the pblock of the out-of-context
+ * kernel.
+ * 
+ */
+public class InlineFlopTools {
+
+    private static final String CLK_OPT = "--clk";
+    private static final String PBLOCK_OPT = "--pblock";
+    private static final String REMOVE_FLOPS_OPT = "--remove_flops";
+
+    public static final String INLINE_SUFFIX = "_rw_inline_flop";
+
+    /**
+     * Add flip flops inline on all the top-level ports of an out-of-context design.
+     * This is useful for placed out-of-context kernels prior to routing so that
+     * after the flops have been placed, the router is forced to route connections
+     * of each of the ports to each of the flops. This can help alleviate congestion
+     * when the kernels are placed/relocated in context. Note this assumes the
+     * design is not implemented as in most contexts it will be placed and routed
```

Comment:
Here it says design is assumed to be not implemented (and there's the assertion to check that it's not placed, below) but on line 71 above it says:

> This is useful for placed out-of-context kernels prior to routing ...

Contradiction?

Reply:
Thanks, I have reworded the comment.